File: Hard/Java/32. Longest Valid Parentheses.java
class Solution {
    public int longestValidParentheses(String s) {
        int open = 0;
        int close = 0;
        int longest = 0;

        for(int i = 0; i < s.length(); i++){
            if(s.charAt(i) == '('){
                open++;
            }else{
                close++;
            }

            if(open == close){
                longest = Math.max(longest, 2 * close);
            }else if(open <= close){
                open = 0;
                close = 0;
            }
        }

        open = 0;
        close = 0;
        
        for(int i = s.length() - 1; 0 <= i; i--){
            if(s.charAt(i) == '('){
                open++;
            }else{
                close++;
            }

            if(open == close){
                longest = Math.max(longest, 2 * open);
            }else if(close <= open){
                open = 0;
                close = 0;
            }
        }

        return longest;
    }
}

File: Hard/Python/32. Longest Valid Parentheses.py
class Solution(object):
    def longestValidParentheses(self, s):
        open_paren = 0
        close_paren = 0
        longest = 0

        for i in range(len(s)):
            if s[i] == '(':
                open_paren += 1
            else:
                close_paren += 1
            
            if open_paren == close_paren:
                longest = max(longest, 2 * close_paren)
            elif open_paren <= close_paren:
                open_paren = 0
                close_paren = 0
        
        open_paren = 0
        close_paren = 0

        for i in range(len(s) - 1, -1, -1):
            if s[i] == '(':
                open_paren += 1
            else:
                close_paren += 1
            
            if open_paren == close_paren:
                longest = max(longest, 2 * open_paren)
            elif close_paren <= open_paren:
                open_paren = 0
                close_paren = 0

        return longest
